### ISSUE TITLE
metadata address search for X86 maybe error

### ISSUE THREAD

**Issue by lhyqy5** [djkaty/Il2CppInspector#1]
In my case.The metadata address  returned in [this code](https://github.com/djkaty/Il2CppInspector/blob/9ced0b6/Il2CppInspector/Il2CppReaderX86.cs#L51-L52) is a pointer in `.got` segment  which  point to real metadata address.

my test file:[testbin.zip](https://github.com/djkaty/Il2CppInspector/files/852369/testbin.zip)

and sorry my poor English.

**Comment by djkaty** [comment]
I'm looking at this now. What are the Unity build target platforms for the files in the 1 and 2 folders (ie. Windows Store, Android etc.)?

**Comment by lhyqy5** [comment]
Only file 1 is built by me on window 10 target to android(NDK version r10e, unity version 5.4.1f1), file 2 is decompressed from this [game](https://play.google.com/store/apps/details?id=jp.co.craftegg.band)

**Comment by Perfare** [comment]
I think it should be like this
<pre>
Image.Position = funcPtr + 0x22;
var ptr = Image.ReadUInt32() + globalOffset;
Image.Position = Image.MapVATR(ptr);
metadata = Image.ReadUInt32();
</pre>

**Comment by djkaty** [comment]
Hey guys,

It's fractionally more complicated than that. In the samples @lhyqy5 supplied, the code which loads the pointer arguments to MetadataCache::Register look like this:
```
.text:0037DAE2                 lea     eax, (unk_4850A4 - 5A229Ch)[ebx]
.text:0037DAE8                 mov     [esp+8], eax
.text:0037DAEC                 mov     eax, ds:(off_5A1D08 - 5A229Ch)[ebx]
.text:0037DAF2                 mov     [esp+4], eax
.text:0037DAF6                 lea     eax, (unk_5ABAD0 - 5A229Ch)[ebx]
.text:0037DAFC                 mov     [esp], eax
```

However the samples I have look like this:
```
.text:005AD704                 lea     eax, (unk_67B375 - 803D70h)[ebx]
.text:005AD70A                 mov     [esp+8], eax
.text:005AD70E                 lea     eax, (unk_802EF4 - 803D70h)[ebx]
.text:005AD714                 mov     [esp+4], eax
.text:005AD718                 lea     eax, (unk_802EA4 - 803D70h)[ebx]
.text:005AD71E                 mov     [esp], eax
```

Notice that the `mov` which loads the 2nd pointer is replaced with a `lea`. The fix is simple of course, I just haven't gotten round to doing it yet - was building a new PC.

Thanks for your efforts in writing the original code Perfare :)

**Comment by lhyqy5** [comment]
Can we  check the meta address if in `.got` segment like this:
```csharp
Image.Position = funcPtr + 0x22;
var ptr = Image.ReadUInt32() + globalOffset;
metadata=ptr;
if(ptr.InSegment(".got")) //need add a function like this
{
    Image.Position = Image.MapVATR(ptr);
    metadata = Image.ReadUInt32();
}
```

**Comment by djkaty** [comment]
Fixed in 0e80351.
